src/datasets/dataset.jl

```diff
@@ -106,13 +106,14 @@ function Base.hcat(x::Vector{<:Real}, d::AbstractDataset{D, T}) where {D, T}
     return Dataset(data)
 end
 
-function Base.hcat(x::AbstractDataset{D1, T}, y::AbstractDataset{D2, T}) where {D1, D2, T}
-    length(x) == length(y) || error("Datasets must be of same length")
-    L = length(x)
-    D = D1 + D2
-    v = Vector{SVector{D, T}}(undef, L)
-    for i = 1:L
-        v[i] = SVector{D, T}((x[i]..., y[i]...,))
+function Base.hcat(ds::Vararg{AbstractDataset{D, T} where {D}, N}) where {T, N}
+    Ls = length.(ds)
+    maxlen = maximum(Ls)
+    all(Ls .== maxlen) || error("Datasets must be of same length")
+    D = sum(dimension.(ds))
```

Comment:
This unecessarily restricts all datasets to have same dimension `D`. Furthermore, it replaces the type parmaeter `D` in the call signature, with a new and different number `D` later on, which is bad practice (because it is typed, otherwise it would be okay).

Given that you never actually utilize the first `D`, you can just make the call signature `AbstractDataset` without `D, T` and get `T = eltype(first(ds))`.